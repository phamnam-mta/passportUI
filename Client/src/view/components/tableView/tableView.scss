.table-view-container {
    max-height: 85vh;
    max-width: 85vw;
    width: auto;
    height: auto;

    .table-title {
        font-size: 24px;
        line-height: 32px;
    }

    .table-scrollable-container {
        display: flex;
        flex-flow: column;
        flex-grow: 1;
        min-height: 0;
        overflow: auto;

        .table-view {
            border-collapse: collapse;
            td,
            th {
                border: 2px solid #8d8d8d;
                max-width: 400px;
                padding: 0.5rem 1rem;
            }
            td:empty::after {
                content: "\00a0";
            }
        }
    }

    .table-view-scrollable-content {
        max-height: 65vh;
    }
}
