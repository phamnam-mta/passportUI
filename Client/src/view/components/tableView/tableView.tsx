import * as React from "react";
import { Stack, IStackTokens } from "@fluentui/react";
import { StudioDocumentTable } from "models/analyzeResult";
import Modal from "../modal/modal";

import "./tableView.scss";

interface ITableViewProps {
    handleTableViewClose: () => any;
    tableToView: StudioDocumentTable;
}

export class TableView extends React.PureComponent<ITableViewProps> {
    private buildTableBody = (): JSX.Element[] | null => {
        const { tableToView: table } = this.props;
        if (!table) {
            return null;
        }

        const stackToken: IStackTokens = {
            childrenGap: 5,
        };
        const tableBody: JSX.Element[] = [];
        const { rowCount, cells } = table;

        for (let i = 0; i < rowCount; i++) {
            const tableRow = [];
            tableBody.push(<tr key={i}>{tableRow}</tr>);
        }

        cells.forEach((cell) => {
            const { rowIndex, columnIndex, rowSpan, columnSpan, kind, content } = cell;

            const cellContentWithSpan = content.map((cellContent, index) => {
                return <span key={index}>{cellContent}</span>;
            });

            tableBody[rowIndex]["props"]["children"][columnIndex] =
                kind === "columnHeader" ? (
                    <th key={columnIndex} colSpan={columnSpan} rowSpan={rowSpan}>
                        <Stack horizontal wrap tokens={stackToken}>
                            {cellContentWithSpan}
                        </Stack>
                    </th>
                ) : (
                    <td key={columnIndex} colSpan={columnSpan} rowSpan={rowSpan}>
                        <Stack horizontal wrap tokens={stackToken}>
                            {cellContentWithSpan}
                        </Stack>
                    </td>
                );
        });

        return tableBody;
    };

    public render() {
        return (
            <Modal
                header={<div className="table-title">Table</div>}
                body={
                    <div className="table-scrollable-container">
                        <table className="table-view">
                            <tbody>{this.buildTableBody()}</tbody>
                        </table>
                    </div>
                }
                footer={null}
                buttonGroup={null}
                onDismiss={this.props.handleTableViewClose}
                onClose={this.props.handleTableViewClose}
                isOpen={this.props.tableToView !== null}
                isDraggable
                containerClassName={"table-view-container"}
                scrollableContentClassName={"table-view-scrollable-content"}
            />
        );
    }
}

export default TableView;
