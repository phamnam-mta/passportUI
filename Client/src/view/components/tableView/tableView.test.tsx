import * as React from "react";
import { shallow } from "enzyme";
import { TableView } from "./tableView";

describe("<TableView />", () => {
    let baseProps;

    beforeEach(() => {
        baseProps = {
            handleTableViewClose: jest.fn(),
            tableToView: {
                rowCount: 3,
                columnCount: 2,
                cells: [
                    {
                        kind: "columnHeader",
                        rowIndex: 0,
                        columnIndex: 0,
                        rowSpan: 1,
                        columnSpan: 1,
                        content: ["PhoneNumber"],
                        boundingRegions: [
                            {
                                pageNumber: 1,
                                boundingBox: [3.3715, 4.8314, 4.4233, 4.8314, 4.4278, 5.0257, 3.3715, 5.0211],
                            },
                        ],
                        spans: [{ offset: 194, length: 11 }],
                    },
                    {
                        kind: "columnHeader",
                        rowIndex: 0,
                        columnIndex: 1,
                        rowSpan: 1,
                        columnSpan: 1,
                        content: ["Name"],
                        boundingRegions: [
                            {
                                pageNumber: 1,
                                boundingBox: [4.4233, 4.8314, 5.3713, 4.8314, 5.3713, 5.0257, 4.4278, 5.0257],
                            },
                        ],
                        spans: [{ offset: 206, length: 4 }],
                    },
                    {
                        rowIndex: 1,
                        columnIndex: 0,
                        rowSpan: 1,
                        columnSpan: 1,
                        content: ["123-456-789"],
                        boundingRegions: [
                            {
                                pageNumber: 1,
                                boundingBox: [3.3715, 5.0211, 4.4278, 5.0257, 4.4278, 5.22, 3.3715, 5.22],
                            },
                        ],
                        spans: [{ offset: 211, length: 11 }],
                    },
                    {
                        rowIndex: 1,
                        columnIndex: 1,
                        rowSpan: 1,
                        columnSpan: 1,
                        content: ["Lee"],
                        boundingRegions: [
                            {
                                pageNumber: 1,
                                boundingBox: [4.4278, 5.0257, 5.3713, 5.0257, 5.3713, 5.22, 4.4278, 5.22],
                            },
                        ],
                        spans: [{ offset: 223, length: 3 }],
                    },
                    {
                        rowIndex: 2,
                        columnIndex: 0,
                        rowSpan: 1,
                        columnSpan: 1,
                        content: ["987-654-321"],
                        boundingRegions: [
                            {
                                pageNumber: 1,
                                boundingBox: [3.3715, 5.22, 4.4278, 5.22, 4.4278, 5.4096, 3.3715, 5.4096],
                            },
                        ],
                        spans: [{ offset: 227, length: 11 }],
                    },
                    {
                        rowIndex: 2,
                        columnIndex: 1,
                        rowSpan: 1,
                        columnSpan: 1,
                        content: ["Wang"],
                        boundingRegions: [
                            {
                                pageNumber: 1,
                                boundingBox: [4.4278, 5.22, 5.3713, 5.22, 5.3713, 5.4096, 4.4278, 5.4096],
                            },
                        ],
                        spans: [{ offset: 239, length: 4 }],
                    },
                ],
                boundingRegions: [
                    { pageNumber: 1, boundingBox: [3.3642, 4.8346, 5.3705, 4.8346, 5.3711, 5.4158, 3.3646, 5.4158] },
                ],
                spans: [{ offset: 194, length: 49 }],
            },
        };
    });

    describe("Rendering", () => {
        it("should match snapshot, when it is open", () => {
            const wrapper = shallow(<TableView {...baseProps} />);

            expect(wrapper).toMatchSnapshot();
        });

        it("should match snapshot, when tableToView is null", () => {
            const props = { ...baseProps, tableToView: null };
            const wrapper = shallow(<TableView {...props} />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should trigger handleTableViewClose, when Modal is closed", () => {
            const wrapper = shallow(<TableView {...baseProps} />);
            const modal = wrapper.find("Modal") as any;

            modal.prop("onClose")();

            expect(baseProps.handleTableViewClose).toBeCalledTimes(1);
        });

        it("should trigger handleTableViewClose, when Modal is dismissed", () => {
            const wrapper = shallow(<TableView {...baseProps} />);
            const modal = wrapper.find("Modal") as any;

            modal.prop("onDismiss")();

            expect(baseProps.handleTableViewClose).toBeCalledTimes(1);
        });
    });
});
