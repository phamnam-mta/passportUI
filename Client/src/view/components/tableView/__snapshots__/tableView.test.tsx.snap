// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableView /> Rendering should match snapshot, when it is open 1`] = `
<Modal
  body={
    <div
      className="table-scrollable-container"
    >
      <table
        className="table-view"
      >
        <tbody>
          <tr>
            <th
              colSpan={1}
              rowSpan={1}
            >
              <Stack
                horizontal={true}
                tokens={
                  Object {
                    "childrenGap": 5,
                  }
                }
                wrap={true}
              >
                <span>
                  PhoneNumber
                </span>
              </Stack>
            </th>
            <th
              colSpan={1}
              rowSpan={1}
            >
              <Stack
                horizontal={true}
                tokens={
                  Object {
                    "childrenGap": 5,
                  }
                }
                wrap={true}
              >
                <span>
                  Name
                </span>
              </Stack>
            </th>
          </tr>
          <tr>
            <td
              colSpan={1}
              rowSpan={1}
            >
              <Stack
                horizontal={true}
                tokens={
                  Object {
                    "childrenGap": 5,
                  }
                }
                wrap={true}
              >
                <span>
                  123-456-789
                </span>
              </Stack>
            </td>
            <td
              colSpan={1}
              rowSpan={1}
            >
              <Stack
                horizontal={true}
                tokens={
                  Object {
                    "childrenGap": 5,
                  }
                }
                wrap={true}
              >
                <span>
                  Lee
                </span>
              </Stack>
            </td>
          </tr>
          <tr>
            <td
              colSpan={1}
              rowSpan={1}
            >
              <Stack
                horizontal={true}
                tokens={
                  Object {
                    "childrenGap": 5,
                  }
                }
                wrap={true}
              >
                <span>
                  987-654-321
                </span>
              </Stack>
            </td>
            <td
              colSpan={1}
              rowSpan={1}
            >
              <Stack
                horizontal={true}
                tokens={
                  Object {
                    "childrenGap": 5,
                  }
                }
                wrap={true}
              >
                <span>
                  Wang
                </span>
              </Stack>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  }
  buttonGroup={null}
  containerClassName="table-view-container"
  footer={null}
  header={
    <div
      className="table-title"
    >
      Table
    </div>
  }
  isDraggable={true}
  isOpen={true}
  onClose={[MockFunction]}
  onDismiss={[MockFunction]}
  scrollableContentClassName="table-view-scrollable-content"
/>
`;

exports[`<TableView /> Rendering should match snapshot, when tableToView is null 1`] = `
<Modal
  body={
    <div
      className="table-scrollable-container"
    >
      <table
        className="table-view"
      >
        <tbody />
      </table>
    </div>
  }
  buttonGroup={null}
  containerClassName="table-view-container"
  footer={null}
  header={
    <div
      className="table-title"
    >
      Table
    </div>
  }
  isDraggable={true}
  isOpen={false}
  onClose={[MockFunction]}
  onDismiss={[MockFunction]}
  scrollableContentClassName="table-view-scrollable-content"
/>
`;
